### PR TITLE
SCEM-205: upgrade webapi

### PR DESCRIPTION
**Pull request by An-Li-magicloud**
Here is the change detail:

- [x] 1. webapi.upload:
    - [x] 1.1 add folder api `_query_or_create_folder(folder_name)`, if name duplication found, it returns the latest folder.
    - [x] 1.2 checking response of s3 uploading, a WebError raised if uploading failure occurs. #223 
    - [x] 1.3 migrate to the new api for task creation.
    - [x] 1.4 print out task url, hence user could click on it and open browser.
- [x] 2. webapi.get_info:
    - [x] 2.1 upgrade to latest api. 
    - [ ] 2.2 After the task is deleted by calling webapi.delete, the Webapi returns 404, instead of a task with deleted status. Hence a better response for querying deleted task is needed.
- [x] 3. webapi.start. 
    - [x] 3.1 upgrade to latest api. 
    - [x] 3.2 removed constant solver version in the config.py and leave solver version variable empty. When solver version has value, protocol version won't be passed to Webapi.
- [x] 4. webapi.delete:
    - [x] 4.1 upgrade to latest api.
- [x] 5. webapi.delete_old: 
    - [x] 5.1 migrate to new endpoint
    - [x] 5.2 add folder name as parameter
- [x] 6. webapi.get_tasks: 
    - [x] 6.1 add folder name as parameter
    - [x] 6.2 migrate to new endpoint

### CODE REVIEW

File: tests/_test_web.py [hunk 1]
```diff
@@ -63,7 +64,7 @@ def test_webapi_2_get_info():
 def test_webapi_3_start():
     """test that we can start running task"""
     task_id = _get_gloabl_task_id()
-    web.start(task_id)
+    web.start(task_id, "release-22.1.5")
```
Comment (lei-flex) [inline tests/_test_web.py:67]: don't hardcode the default version.

File: tests/_test_web.py [hunk 2]
```diff
@@ -88,12 +89,12 @@ def test_webapi_6_load():
...
 
-def _test_webapi_7_delete():
+def test_webapi_7_delete():
     """test that we can monitor task"""
     task_id = _get_gloabl_task_id()
     web.delete(task_id)
-    task_info = web.get_info(task_id)
-    assert task_info.status in ("deleted", "deleting")
+    # task_info = web.get_info("9f953cc3-f0bd-49e7-b90d-3161b1ed1187")
+    # assert not task_info
```
Comment (lei-flex) [inline tests/_test_web.py:97]: why removed?

File: tidy3d/web/config.py [hunk 3]
```diff
@@ -27,7 +27,8 @@ class WebConfig:  # pylint:disable=too-many-instance-attributes
     s3_region="us-east-1",
     studio_bucket="flow360-studio-v1",
     auth_api_endpoint="https://portal-api.dev-simulation.cloud",
-    web_api_endpoint="https://tidy3d-api.dev-simulation.cloud",
+    # web_api_endpoint="https://tidy3d-api.dev-simulation.cloud",
+    web_api_endpoint="http://localhost:5000",
```
Comment (lei-flex) [inline tidy3d/web/config.py:31]: change back to dev url

File: tidy3d/web/httputils.py [hunk 4]
```diff
@@ -39,8 +39,8 @@ def wrapper(*args, **kwargs):
 
         # if it was successful, try returning data from the response
         try:
-            json_data = resp.json()["data"]
-            return json_data
+            json_str = resp.json()
+            return json_str["data"] if "data" in json_str else json_str
```
Comment (lei-flex) [inline tidy3d/web/httputils.py:43]: the webapi should be updated to follow the json format.

File: tidy3d/web/webapi.py [hunk 5]
```diff
@@ -110,43 +111,52 @@ def upload(
...
-def get_info(task_id: TaskId) -> TaskInfo:
+def get_info(task_id: TaskId, folder_name: str = "default") -> TaskInfo:
     """Return information about a task.
 
     Parameters
     ----------
     task_id : str
         Unique identifier of task on server.  Returned by :meth:`upload`.
+    folder_name: str
+        get task under certain folder.
```
Comment (lei-flex) [inline tidy3d/web/webapi.py:122]: change to: the folder name containing the task.

File: tidy3d/web/webapi.py [hunk 6]
```diff
@@ -337,7 +347,7 @@ def delete(task_id: TaskId) -> TaskInfo:
         Object containing information about status, size, credits of task.
     """
 
-    method = f"fdtd/task/{str(task_id)}"
+    method = f"tidy3d/projects/any/tasks/{task_id}"
```
Comment (lei-flex) [inline tidy3d/web/webapi.py:350]: let's do it correctly.  the webapi looks strange.
use delete f"tidy3d/tasks/{task_id}/"

File: tidy3d/web/webapi.py [hunk 7]
```diff
@@ -418,28 +428,38 @@ def get_tasks(num_tasks: int = None, order: Literal["new", "old"] = "new") -> Li
     return out_dict
 
 
+def _query_or_create_folder(folder_name) -> Folder:
```
Comment (lei-flex) [inline tidy3d/web/webapi.py:431]: what if  multiple thread submit same folder name?

File: tidy3d/web/webapi.py [hunk 8]
```diff
@@ -418,28 +428,38 @@ def get_tasks(num_tasks: int = None, order: Literal["new", "old"] = "new") -> Li
...
+        callback_url: str = None,
 ) -> TaskId:
     """upload with all kwargs exposed"""
 
+    folder = _query_or_create_folder(folder_name)
     simulation.validate_contents()
 
     json_string = simulation._json_string()  # pylint:disable=protected-access
     data = {
-        "status": "uploading",
```
Comment (lei-flex) [inline tidy3d/web/webapi.py:435]: nice change!

File: tidy3d/web/webapi.py [hunk 9]
```diff
@@ -461,15 +481,18 @@ def _upload_task(  # pylint:disable=too-many-locals,too-many-arguments
...
+    resp = client.put_object(
         Body=json_string,
         Bucket=bucket,
         Key=key,
         # Callback=upload_progress.report
     )
 
-    task["status"] = "draft"
-    http.put(f"{method}/{task_id}", data=task)
+    if resp.get("ResponseMetadata") and resp.get("ResponseMetadata").get("HTTPStatusCode") != 200:
```
Comment (lei-flex) [inline tidy3d/web/webapi.py:491]: nice checking here.

File: tidy3d/web/webapi.py [hunk 10]
```diff
@@ -461,15 +481,18 @@ def _upload_task(  # pylint:disable=too-many-locals,too-many-arguments
...
+        raise WebError(f"Upload file to 3S failed, please check your network or try it later.")
+
+    # task["status"] = "draft"
```
Comment (lei-flex) [inline tidy3d/web/webapi.py:494]: @momchil-flex could u check if we still need draft status from python client side?

Comment (lei-flex) [reply tidy3d/web/webapi.py:494]: if it is just medium status, we can remove the code.

**Comment by tylerflex** [comment]
Just FYI and for future reference, the tests will not pass unless the following pass (run from root of this directory)
`black . --line-length 100` gives no output (no files changed).
`python lint.py` gives 10/10 score (no linting errors).

File: tests/_test_web.py (hunk 1 above)
Comment (An-Li-magicloud) [reply tests/_test_web.py:67]: It is for missing mapping.

File: tests/_test_web.py (hunk 2 above)
Comment (An-Li-magicloud) [reply tests/_test_web.py:97]: Deleted task query will return 404.

**Comment by tylerflex** [comment]
@An-Li-magicloud Can you provide some description for these changes in the PR?  What is the goal and effects of them?

**Comment by An-Li-magicloud** [comment]
@tylerflex do you think we still need point 5 and 6?

**Comment by momchil-flex** [comment]
We need 5, because we still don't have a very good way for users to delete a large number of jobs. Unless this will be something that will soon be available in the web client? I'm thinking of file manager-like multiple selection of simulations.

**Comment by tylerflex** [comment]
Yea I think 5 and 6 are still needed like momchil said.  There's no better way to delete or select large amounts of task data.

File: tidy3d/web/webapi.py [hunk 11]
```diff
@@ -432,14 +450,12 @@ def _upload_task(  # pylint:disable=too-many-locals,too-many-arguments
 
     json_string = simulation._json_string()  # pylint:disable=protected-access
     data = {
-        "status": "uploading",
-        "solverVersion": solver_version,
         "taskName": task_name,
-        "workerGroup": worker_group,
```
Comment (momchil-flex) [inline tidy3d/web/webapi.py:438]: @An-Li-magicloud this tells the resource allocation daemon which worker to use.

**Pull request merged**